library/MiscUtils.cpp

```diff
@@ -83,6 +85,45 @@ DFHACK_EXPORT void imbue_with_locale(std::ostringstream &ss, NumberFormatType ty
     }
     if (type != NumberFormatType::SCIENTIFIC)
         ss << std::fixed;
+    ss << std::setprecision(2);
+}
+
+// magnitude to (suffix, divisor)
+static const std::vector<std::pair<const char *, double>> sig_fig_db = {
+    { "", 1.0},              //                    0.*
+    { "", 1.0},              //                    1
+    { "", 1.0},              //                   10
+    { "", 1.0},              //                  100
+    { "", 1.0},              //                1,000
+    {"k", 1000.0},           //               10,000
+    {"k", 1000.0},           //              100,000
+    {"M", 1000000.0},        //            1,000,000
+    {"M", 1000000.0},        //           10,000,000
+    {"M", 1000000.0},        //          100,000,000
+    {"B", 1000000000.0},     //        1,000,000,000
+    {"B", 1000000000.0},     //       10,000,000,000
+    {"B", 1000000000.0},     //      100,000,000,000
+    {"T", 1000000000000.0},  //    1,000,000,000,000
+    {"T", 1000000000000.0},  //   10,000,000,000,000
+    {"T", 1000000000000.0},  //  100,000,000,000,000+
+};
+
+DFHACK_EXPORT std::string format_number_by_sig_fig(double num, size_t sig_figs) {
+    if (num == 0)
+        return "0";
+
+    std::ostringstream ss;
+    if (sig_figs > sig_fig_db.size()) {
+        ss << std::fixed << num;
+        return ss.str();
+    }
+
+    int sf = (int)sig_figs;
+    int magnitude = std::min((size_t)std::floor(std::log10(std::abs(num))) + 1, sig_fig_db.size() - 1);
```

Comment:
this code will behave very strangely for arguments between -1.0 and 1.0 (except for 0.0 exactly): log10 will be negative, which, when cast to the unsigned `size_t`, convert to an extremely large positive number, which at least gets clamped by `std::min` to the highest prefix. as a result, arguments between -1.0 and 1.0 (excluding 0) will result in the `T` prefix being used on the unit... which is almost certainly not at all what is intended or desired